redsky/event_streams.py

```diff
@@ -0,0 +1,711 @@
+        created, but a stop document will be issued.
+        """
+        if not self.event_failed:
+            if self.run_start_uid is None:
+                raise RuntimeError("Received Event before RunStart.")
+            if isinstance(outputs, Exception):
+                return self.stop(outputs)
+
+            # Make a new event with no data
+            if len(self.output_info) == 1:
+                outputs = (outputs,)
+
+            new_event = dict(uid=str(uuid.uuid4()),
+                             time=time.time(),
+                             timestamps={},
+                             descriptor=self.outbound_descriptor_uid,
+                             filled={k[0]: True for k in self.output_info},
+                             seq_num=self.i)
+
+            if self.output_info:
+                new_event.update(data={output_name: output
+                                       for (output_name, desc), output in
+                                       zzip(self.output_info, outputs)})
+            else:
+                new_event.update(data=outputs['data'])
+            self.i += 1
+            return new_event
+
+    def refresh_event(self, event):
+        """Issue a new event
+
+        Parameters
+        ----------
+        event: tuple, dict, or other
+            The event document
+        Returns
+        -------
+        new_event: dict
+            A new event with the same core data
+        """
+        if not self.event_failed:
+            if self.run_start_uid is None:
+                raise RuntimeError("Received Event before RunStart.")
+            if isinstance(event, Exception):
+                return self.stop(event)
+
+            new_event = dict(event)
+            new_event.update(dict(uid=str(uuid.uuid4()),
+                                  time=time.time(),
+                                  timestamps={},
+                                  seq_num=self.i))
+
+            self.i += 1
+            return new_event
+
+
+class map(EventStream):
+    """Apply a function onto every event in the stream"""
+    def __init__(self, func, child, *,
+                 output_info=None, input_info=None,
+                 **kwargs):
+        """Initialize the node
+
+        Parameters
+        ----------
+        func: callable
+            The function to map on the event data
+        child: EventStream instance
+            The source of the data
+        input_info: dict
+            describe the incoming streams
+        output_info: list of tuples
+            describe the resulting stream
+        """
+        self.func = func
+        self.kwargs = kwargs
+
+        EventStream.__init__(self, child, output_info=output_info,
+                             input_info=input_info, **kwargs)
+        self.generate_provenance(func)
+
+    def event(self, docs):
+        try:
+            # we need to expose the raw event data
+            res = self.event_guts(docs)
```

Comment:
what if you wanted to map a function that explicitly handled the event doc? could be worth considering a `raw` flag here, maybe in the future.